### PR TITLE
tests: init RobotFramework and add basic xtimer test

### PR DESCRIPTION
**Pull request by smlng**
### Contribution description

This introduces common files to enable RobotFramework based (HIL) tests in RIOT.
It also ships a very basic xtimer test application utilising the RobotFramework.
The latter will be extended later on, but should give others a starting point to see how to
use RF with RIOT.

RF tests (currently) run from a distinct make target, i.e. `robot-test`; so will not interfere with existing tests. This might be changed later on, but for now it should ease introduction of RF tests w/o breaking existing stuff.

### Testing procedure

Currently this only works for real boards connected via serial, so connect you favourite board and run

```
BOARD=<name> make -C tests/xtimer_cli flash robot-test
```

### Issues/PRs references

- replaces #10095
- partly implements stuff from #10624

### CODE REVIEW

File: tests/xtimer_cli/README.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Xtimer API test
+
+This application enables tests of xtimer API calls using the RIOT shell.
+Consult the 'help' shell command for available actions.
+
+## Requirements
+
+The automated tests are based on RobotFramework and thus depend on additional
+Python packages, namely `robotframework` and `riot_pal`.
```
Comment (MichelRottleuthner) [inline tests/xtimer_cli/README.md]: maybe add a reference to dist/robotframework/README.md for installing the requirements here?

File: tests/xtimer_cli/main.c [hunk 2]
```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (C) 2018 HAW Hamburg
```
Comment (MichelRottleuthner) [inline tests/xtimer_cli/main.c]: 2018? ;)

File: tests/xtimer_cli/main.c [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+int cmd_get_metadata(int argc, char **argv)
+{
+    (void)argv;
+    (void)argc;
+
+    printf("Success: [%s, %s]\n", RIOT_BOARD, RIOT_APPLICATION);
+
+    return 0;
+}
+
```
Comment (MichelRottleuthner) [inline tests/xtimer_cli/main.c]: Didn't look it up, but is the `echo` cmd described in #10624 already used by riot_pal? If yes, maybe add it here?

File: tests/xtimer_cli/tests/01__xtimer_base.robot [hunk 4]
```diff
@@ -0,0 +1,27 @@
+*** Settings ***
+Documentation       Basic tests to verify functionality of the Xtimer API.
+
+Suite Setup         Run Keywords    Reset Application
+...                                 DUT Must Have API Firmware
+Test Setup          Run Keywords    Reset Application
+...                                 DUT Must Have API Firmware
```
Comment (MichelRottleuthner) [inline tests/xtimer_cli/tests/01__xtimer_base.robot]: As this is somehow a POC/showcase for using robot framework a little more comments here would be nice to make clear what `Suite Setup`, `Test Setup` and so on actually does. 
Also IIRC the second `DUT Must Have API Firmware` is a little hack to ensure the communication channel is working correctly. -> this should be noted explicitly to make it clear to others that this is required.

**Comment by tcschmidt** [comment]
> 
> 
> Blocking to avoid unexpected merge.

What is it you want to question or change?

File: tests/xtimer_cli/README.md (hunk 1 above)
Comment (smlng) [reply tests/xtimer_cli/README.md]: yep makes sense

File: tests/xtimer_cli/main.c (hunk 2 above)
Comment (smlng) [reply tests/xtimer_cli/main.c]: it's an old one

File: tests/xtimer_cli/main.c (hunk 3 above)
Comment (smlng) [reply tests/xtimer_cli/main.c]: nope, no echo there, and I would rather make (as discussed offline) a PR introducing both (metadata and echo) as a shell module, and not start adding stuff here that is not needed. I need the metadata but I don't need or use `echo` right now, so why bother.

File: tests/xtimer_cli/tests/01__xtimer_base.robot (hunk 4 above)
Comment (smlng) [reply tests/xtimer_cli/tests/01__xtimer_base.robot]: first: why should I repeat the RF documentation? There is plenty documentation on Suite/Test Setup at the original sources, I can add some links in the dist/robotframework/README.md but copy-pasting docu doesn't make sense (to me).

second: no that's no hack in that sense, it's just there to ensure that before each test the devices are reset and responding correctly before running the tests. The Suite Setup could be avoided but on the other it helps to skip all tests if there is a problem with the board.

**Comment by smlng** [comment]
addressed comments by @MichelRottleuthner

**Comment by cladmi** [comment]
This fails for me:

```
pip install --upgrade --user -r dist/robotframework/requirements.txt
Collecting robotframework (from -r dist/robotframework/requirements.txt (line 1))
  Using cached https://files.pythonhosted.org/packages/36/c6/6f89c80ac5a526a091bd383ffdfc64c9a68d9df0c775d4b36f03d8e0ac25/robotframework-3.1.1-py2.py3-none-any.whl
Collecting riot_pal==0.2.2 (from -r dist/robotframework/requirements.txt (line 2))
  Could not find a version that satisfies the requirement riot_pal==0.2.2 (from -r dist/robotframework/requirements.txt (line 2)) (from versions: )
No matching distribution found for riot_pal==0.2.2 (from -r dist/robotframework/requirements.txt (line 2))
```

**Comment by cladmi** [comment]
As our target is ubuntu stable, `pip` and `python` are by default `python2` so the doc should match this.


> Currently this only works for real boards connected via serial, so connect you favourite board and run

So it does not work with `native` but also does not work with any board using `stdio_rtt`.
So no `ruuvitag`, `hamilton` (pr #9013 ), `thingy52`.

Please do no re-invent the wheel and use `make term` as a base.
Looking on how it is handled, I am pretty sure it does not handle a `thingy52` a `ruuvitag` or the

**Comment by tcschmidt** [comment]
> Since I'm not aware of any conclusion with #10241 (where the discussion is not going toward the direction of using the Robot Framework IMHO), I don't want to have this merged by mistake before the next release.

It's feature-freeze now, so no danger to sneak into the release: You may want to remove the operational block then @aabadie ?

**Comment by aabadie** [comment]
> so no danger to sneak into the release: You may want to remove the operational block

Sorry, but no. My main blocker point is still not full-filled: there's no consensus whether we add and maintain RF files (and all the custom keywords that will come with) in the RIOT repository or not.

**Comment by smlng** [comment]
> It is going in a direction to put the abstraction code as a RobotFramework specific class https://github.com/RIOT-OS/RIOT/pull/10843/files#diff-2756e48a9944cf93cbea221c7c933574
> So forces into this scheme anyway.

That small python wrapper for Xtimer is RF related indeed, at least the following lines are required to handle the imports correctly and auto-generate keywords in RF:
```
    ROBOT_LIBRARY_SCOPE = 'TEST SUITE'
    ROBOT_LIBRARY_VERSION = get_version()
```

However, this does not mean this class/lib cannot be used by another framework. Is there a (simple) way to make the imports and those two variables conditional, i.e. only have them when this file is used with RF?

> 
> And it also does not reuse the normal RIOT way to access the serial port which is `make term` and so fails with `native`, IoT-LAB integration (or any node access through ssh like the Hamburg HIL nodes) and boards that cannot be accessed directly through only a serial. So it does not interact correctly with RIOT and is really specific to your own setup.
> 
I agree, and I'm working on changing this to work with `make term` and thus also allow to use it with IoT-Lab.

**Comment by tcschmidt** [comment]
See related PR: RIOT-OS/riot_pal#24

File: tests/xtimer_cli/tests/Xtimer.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
+# Copyright (C) 2018 Kevin Weiss <kevin.weiss@haw-hamburg.de>
```
Comment (MrKevinWeiss) [inline tests/xtimer_cli/tests/Xtimer.py]: Ahh this is what thomas was talking about, please change to Copyright (C) 2018 Kevin Weiss, for HAW <kevin.weiss@haw-hamburg.de> (or remove completely).

File: tests/xtimer_cli/tests/Xtimer.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+"""@package PyToAPI
+This module handles parsing of information from RIOT xtimer_cli test.
+"""
+import logging
+
+from riot_pal import DutShell
+from robot.version import get_version
```
Comment (MrKevinWeiss) [inline tests/xtimer_cli/tests/Xtimer.py]: Also this should be separated out into a robot independent interface then a simple wrapper for robot.

File: dist/robotframework/lib/PhilipAPI.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
+from riot_pal import LLMemMapIf, PHILIP_MEM_MAP_PATH
+from robot.version import get_version
+from time import sleep
+
+class PhilipAPI(LLMemMapIf):
```
Comment (MrKevinWeiss) [inline dist/robotframework/lib/PhilipAPI.py]: Maybe we start without philip at all and just do non-philip tests to show as an example.

File: dist/robotframework/res/philip.keywords.txt [hunk 8]
```diff
@@ -0,0 +1,11 @@
+*** Settings ***
+Library             PhilipAPI  port=%{PHILIP_PORT}  baudrate=${115200}  WITH NAME  PHILIP
```
Comment (MrKevinWeiss) [inline dist/robotframework/res/philip.keywords.txt]: maybe remove for now...

**Comment by smlng** [comment]
closing as this moved to separate repo for faster integration and progression, see https://github.com/RIOT-OS/RobotFW-tests

**Pull request closed without merging**